README.md
# file-cache

<p align="center">
  <img src="https://cdn.yiranzai.cn/yiranzai/logo/mouse/mouse1.png" alt="" width="20%">
</p>

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

PHP File Cache, use [Separate Chaining](https://en.wikipedia.org/wiki/Hash_table#separate_chaining) resolve hash conflict. one cache, one file, one bucket.

## Feature

-   Cache can set expired time.
-   Use [Separate Chaining](https://en.wikipedia.org/wiki/Hash_table#separate_chaining) resolve hash conflict.
-   Hash key is the path to store data.

## Structure

If any of the following are applicable to your project, then the directory structure should follow industry best practices by being named the following.

```
src/
tests/
vendor/
```

## Install

Via Composer

```bash
$ composer require yiranzai/file-cache
```

## Usage

### API

This package provides these methods.

-   `put($key, $data, $minutes): Cache`  put one cache to file,set expired time
    -   `$minutes` can be a `int` or `DateTime` or `null` or [Supported Date and Time Formats ](http://us1.php.net/manual/zh/datetime.formats.php)
-   `forever($key, $data): Cache`  forever save one cache to file
-   `get($key, $default = null): ?string`  get the data corresponding to the key
-   `delete($key): bool`  delete one cache
-   `flush(): void`  delete all cache
-   `dataPath($path): self`  change data save path

### Demo

```php
$cache = new Yiranzai\File\Cache();

$cache->put('key', 'data', 10);
$cache->put('key1', 'data1', new DateTime());
$cache->put('key2', 'data2', 'now');
$cache->forever('key3', 'data3');

$cache->get('key');   // data

$cache->delete('key');   // true

$cache->flush();

$cache->get('not_exists','nothing');   // nothing
$cache->dataPath('YOUR_PATH');

// or

$cache = new Yiranzai\File\Cache(['dataPath'=>'YOUR_PATH']);
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

```bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email wuqingdzx@gmail.com instead of using the issue tracker.

## Credits

-   [yiranzai][link-author]
-   [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/yiranzai/file-cache.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/yiranzai/php-file-cache/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/yiranzai/php-file-cache.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/yiranzai/php-file-cache.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/yiranzai/file-cache.svg?style=flat-square
[link-packagist]: https://packagist.org/packages/yiranzai/file-cache
[link-travis]: https://travis-ci.org/yiranzai/php-file-cache
[link-scrutinizer]: https://scrutinizer-ci.com/g/yiranzai/php-file-cache/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/yiranzai/php-file-cache
[link-downloads]: https://packagist.org/packages/yiranzai/file-cache
[link-author]: https://github.com/yiranzai
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `file-cache` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## [1.0.0] - 2019-03-14

### Added

-   `forever($key, $data): Cache`  forever save one cache to file
-   `delete($key): bool`  delete one cache
-   `flush(): void`  delete all cache

### Deprecated

-   `put($key, $data, $minutes): Cache`  put one cache to file,set expired time
    -   `$minutes` can be a `int` or `DateTime` or `null` or [Supported Date and Time Formats ](http:us1.php.net/manual/zh/datetime.formats.php)
-   `get($key, $default = null): ?string`  get the data corresponding to the key
-   `dataPath($path): self`  change data save path

### Fixed

-   Bucket::delNode() delete head has bug

### Removed

-   Nothing

### Security

-   Nothing


tests/CacheTest.php
<?php

declare(strict_types=1);

namespace Yiranzai\File\Tests;

use PHPUnit\Framework\TestCase;
use Yiranzai\File\Cache;
use Yiranzai\File\Filesystem;

class CacheTest extends TestCase
{
    /**
     * @throws \Exception
     */
    public function testCacheDefault()
    {
        $str      = 'test';
        $path     = __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR;
        $dataPath = __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . 'data2' . DIRECTORY_SEPARATOR;
        $file     = new Filesystem();
        $cache    = new Cache(['file' => 'tests', $str => $str]);
        $cache->put($str, $str);
        $this->assertSame($str, $cache->get($str));
        $this->assertSame('abc', $cache->get('abc', 'abc'));

        $this->assertTrue($file->exists($path));
        $file->deleteDirectory($path);
        $this->assertFalse($file->exists($path));

        $cache->dataPath($dataPath);
        $cache->put($str, $str);
        $this->assertSame($str, $cache->get($str));
        $this->assertSame('abc', $cache->get('abc', 'abc'));

        $this->assertTrue($file->exists($dataPath));
        $file->deleteDirectory($dataPath);
        $this->assertFalse($file->exists($dataPath));
    }

    /**
     * @throws \Exception
     */
    public function testDelete()
    {
        $path   = __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR;
        $strOne = 'test1';
        $cache  = new Cache();
        $file   = new Filesystem();
        $cache->put($strOne, $strOne);
        $this->assertTrue($cache->delete($strOne));
        $this->assertNull($cache->get($strOne));
        $this->assertTrue($cache->delete('nothing'));

        $this->assertTrue($file->exists($path));
        $file->deleteDirectory($path);
        $this->assertFalse($file->exists($path));
    }

    /**
     * @throws \Exception
     */
    public function testFlush()
    {
        $path     = __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR;
        $strOne   = 'test1';
        $strTwo   = 'test2';
        $strThree = 'test3';
        $strFour  = 'test4';
        $cache    = new Cache();
        $file     = new Filesystem();
        $date     = new \DateTime();
        $date->sub(new \DateInterval('PT10M'));
        $cache->forever($strOne, $strOne);
        $cache->put($strTwo, $strTwo, 10);

        $cache->put($strThree, $strThree, $date);
        $this->assertNull($cache->get($strThree));

        $cache->put($strFour, $strFour, 'now');

        $cache->flush();

        $this->assertFalse($file->exists($path));
    }

    public function testDateExcetion()
    {
        $this->expectException(\Exception::class);
        $strOne = 'test1';

        $cache = new Cache();
        $cache->put($strOne, $strOne, 'asdhadjksh');
    }
}


tests/BucketTest.php
<?php

declare(strict_types=1);

namespace Yiranzai\File\Tests;

use PHPUnit\Framework\TestCase;
use Yiranzai\File\Bucket;

class BucketTest extends TestCase
{
    public function testBucket()
    {
        $strOne   = 'test1';
        $strTwo   = 'test2';
        $strThree = 'test3';
        $strFour  = 'test4';
        $strFive  = 'test5';
        $bucket   = new Bucket($strOne, $strOne);
        $bucket->putNode($strTwo, $strTwo);
        $this->assertSame(2, $bucket->countNode());
        $this->assertSame($strTwo, $bucket->findNode($strTwo)->data);
        $this->assertSame($strOne, $bucket->findNode($strOne)->data);
        $bucket->putNode($strOne, $strTwo);
        $this->assertSame($strTwo, $bucket->findNode($strOne)->data);
        $bucket->putNode($strTwo, $strOne);
        $this->assertSame($strOne, $bucket->findNode($strTwo)->data);
        $this->assertNull($bucket->findNode('nothing'));
        $bucket->putNode($strOne, $strOne);
        $bucket->putNode($strTwo, $strTwo);
        $bucket->addNode($strThree, $strThree);
        $bucket->addNode($strFive, $strFive);
        $bucket->insertNode($strFour, $strThree, $strFour);
        $bucket->showNode();
        $this->assertSame($bucket->findNode($strFour)->next->data, $bucket->find($strFive));
        $bucket->delNode($strFour);
        $this->assertNull($bucket->findNode($strFour));
        $bucket->delNode($strOne);
        $this->assertNull($bucket->findNode($strOne));
    }
}


tests/FilesystemTest.php
<?php

declare(strict_types=1);

namespace Yiranzai\File\Tests;

use Yiranzai\File\Filesystem;
use PHPUnit\Framework\TestCase;

class FilesystemTest extends TestCase
{
    public function testFile()
    {
        $file      = new Filesystem();
        $str       = 'test';
        $path      = __DIR__ . DIRECTORY_SEPARATOR . 'testOne.txt';
        $pathTwo   = __DIR__ . DIRECTORY_SEPARATOR . 'testTwo.yaml';
        $pathThree = __DIR__ . DIRECTORY_SEPARATOR . 'testThree.ini';
        $file->prepend($path, $str);
        $this->assertSame($file->basename($path), pathinfo($path, PATHINFO_BASENAME));
        $this->assertSame($file->name($path), pathinfo($path, PATHINFO_FILENAME));
        $this->assertSame($file->dirname($path), pathinfo($path, PATHINFO_DIRNAME));
        $this->assertSame($file->extension($path), pathinfo($path, PATHINFO_EXTENSION));
        $this->assertSame($file->type($path), filetype($path));
        $this->assertSame($file->hash($path), md5_file($path));
        $this->assertSame($file->mimeType($path), finfo_file(finfo_open(FILEINFO_MIME_TYPE), $path));
        $this->assertTrue($file->isReadable($path));
        $this->assertTrue($file->isWritable($path));
        $file->chmod($path, 0000);
        $this->assertFalse($file->isReadable($path));
        $this->assertFalse($file->isWritable($path));
        $file->chmod($path, 0755);
        $this->assertSame($file->lastModified($path), filemtime($path));
        $this->assertTrue($file->exists($path));
        $this->assertSame($str, $file->get($path));

        $file->copy($path, $pathTwo);
        $this->assertTrue($file->exists($path));
        $this->assertTrue($file->exists($pathTwo));
        $this->assertSame($str, $file->get($pathTwo));
        $this->assertSame($str, $file->sharedGet($pathTwo));
        $file->append($path, $str . '1');
        $this->assertSame($str . $str . '1', $file->get($path));
        $file->prepend($path, $str . '2');
        $this->assertSame($str . '2' . $str . $str . '1', $file->get($path));

        $file->delete($path);
        $this->assertFalse($file->exists($path));

        $file->move($pathTwo, $pathThree);
        $this->assertFalse($file->exists($pathTwo));
        $this->assertTrue($file->exists($pathThree));
        $this->assertSame($str, $file->get($pathThree));
        $this->assertSame($str, $file->sharedGet($pathThree));

        $file->delete($pathThree);
        $this->assertFalse($file->exists($pathThree));
    }

    public function testDir()
    {
        $file      = new Filesystem();
        $path      = __DIR__ . DIRECTORY_SEPARATOR . 'testOne';
        $children  = ['aaa', 'bbb', 'ccc'];
        $pathTwo   = __DIR__ . DIRECTORY_SEPARATOR . 'testTwo';
        $pathThree = __DIR__ . DIRECTORY_SEPARATOR . 'testThree';
        $file->makeDirectory($path);
        foreach ($children as $child) {
            $file->makeDirectory($path . DIRECTORY_SEPARATOR . $child);
        }
        $this->assertTrue($file->exists($path));
        foreach ($children as $child) {
            $this->assertTrue($file->exists($path . DIRECTORY_SEPARATOR . $child));
        }
        $this->assertTrue($file->isDirectory($path));
        $this->assertTrue($file->isWritable($path));
        $file->chmod($path, 0000);
        $this->assertFalse($file->isReadable($path));
        $this->assertFalse($file->isWritable($path));
        $file->chmod($path, 0755);


        $file->copyDirectory($path, $pathTwo);
        $this->assertTrue($file->exists($path));
        $this->assertTrue($file->exists($pathTwo));

        $file->deleteDirectory($path);
        $this->assertFalse($file->exists($path));

        $file->moveDirectory($pathTwo, $pathThree);
        $this->assertFalse($file->exists($pathTwo));
        $this->assertTrue($file->exists($pathThree));

        $file->deleteDirectory($pathThree);
        $this->assertFalse($file->exists($pathThree));
    }

    public function testDirTwo()
    {
        $file      = new Filesystem();
        $path      = __DIR__ . DIRECTORY_SEPARATOR . 'testOne';
        $pathTwo   = __DIR__ . DIRECTORY_SEPARATOR . 'testTwo';
        $pathThree = __DIR__ . DIRECTORY_SEPARATOR . 'testThree';
        $file->makeDirectory($path, 0755, false, true);
        $this->assertTrue($file->isDirectory($path));
        $this->assertTrue($file->isWritable($path));
        $file->chmod($path, 0000);
        $this->assertFalse($file->isReadable($path));
        $this->assertFalse($file->isWritable($path));
        $file->chmod($path, 0755);
        $this->assertTrue($file->exists($path));

        $file->copyDirectory($path, $pathTwo);
        $this->assertTrue($file->exists($path));
        $this->assertTrue($file->exists($pathTwo));
        $this->assertSame(substr(sprintf('%o', fileperms($path)), -4), $file->chmod($path));

        $file->cleanDirectory($path);
        $this->assertTrue($file->exists($path));
        $file->deleteDirectory($path);
        $this->assertFalse($file->exists($path));

        $file->moveDirectory($pathTwo, $pathThree);
        $this->assertFalse($file->exists($pathTwo));
        $this->assertTrue($file->exists($pathThree));
        $this->assertFalse($file->moveDirectory($pathTwo, $pathThree, true));
        $this->assertFalse($file->exists($pathThree));

        $file->deleteDirectory($pathThree);
        $this->assertFalse($file->exists($pathThree));

        $this->assertSame($file->windowsOs(), stripos(PHP_OS, 'win') === 0);
    }

    public function testDirException()
    {
        $this->expectException(\Exception::class);
        $file = new Filesystem();
        $file->get(__DIR__);
    }
}


src/Node.php
<?php

namespace Yiranzai\File;

use DateTime;

/**
 * Class Node
 * @package Yiranzai\Dht
 */
class Node
{
    /**
     * @var string
     */
    public $data;
    /**
     * @var Node|null
     */
    public $next;
    /**
     * @var string
     */
    public $key;
    /**
     * @var DateTime
     */
    protected $expiredAt;

    /**
     * Node constructor.
     * @param string   $key
     * @param string   $data
     * @param DateTime $expiredAt
     */
    public function __construct(string $key, string $data, DateTime $expiredAt = null)
    {
        $this->key       = $key;
        $this->data      = $data;
        $this->expiredAt = $expiredAt;
    }

    /**
     * @param null|string $default
     * @return string|null
     * @throws \Exception
     */
    public function getData($default = null): ?string
    {
        if ($this->isExpired()) {
            return $default;
        }
        return $this->data;
    }

    /**
     * @return bool
     * @throws \Exception
     */
    public function isExpired(): bool
    {
        if ($this->expiredAt === null) {
            return false;
        }
        return $this->expiredAt < new DateTime();
    }
}


src/Cache.php
<?php

namespace Yiranzai\File;

use DateTime;
use DateInterval;

/**
 * Class Cache
 * @package Yiranzai\File
 */
class Cache
{
    public const DEFAULT_PATH = __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . 'data'
    . DIRECTORY_SEPARATOR;
    /**
     * @var string
     */
    protected $dataPath = self::DEFAULT_PATH;
    /**
     * @var Filesystem
     */
    protected $file;
    /**
     * @var array
     */
    protected $guarded = ['file'];

    /**
     * Hash constructor.
     * @param array $config
     */
    public function __construct($config = [])
    {
        if (!empty($config)) {
            foreach ($config as $key => $item) {
                if (in_array($key, $this->guarded, true)) {
                    continue;
                }
                $this->$key = $item;
            }
        }
        $this->file = new Filesystem();
    }

    /**
     * 存储数据
     *
     * @param string                   $key
     * @param string                   $data
     * @param string|DateTime|int|null $minutes
     * @return Cache
     * @throws \Exception
     */
    public function put(string $key, string $data, $minutes = null): Cache
    {
        $this->ensureCacheDirectoryExists($path = $this->path($key));

        if ($this->file->exists($path = $this->path($key))) {
            if ($bucket = $this->getBucket($path)) {
                $bucket->putNode($key, $data, $this->generateDate($minutes));
            }
        } else {
            $bucket = new Bucket($key, $data, $this->generateDate($minutes));
        }
        $this->file->put($path, serialize($bucket));
        return $this;
    }

    /**
     * @param string $key
     * @param string $data
     * @return Cache
     * @throws \Exception
     */
    public function forever(string $key, string $data): Cache
    {
        $this->put($key, $data);
        return $this;
    }

    /**
     * @param $minutes
     * @return DateTime
     * @throws \Exception
     */
    protected function generateDate($minutes): ?DateTime
    {
        if ($minutes === null) {
            return null;
        }
        if ($minutes instanceof \DateTime) {
            return $minutes;
        }
        if (is_int($minutes)) {
            $date = new DateTime();
            return $date->add(new DateInterval('PT' . $minutes . 'M'));
        }
        try {
            return new DateTime($minutes);
        } catch (\Exception $exception) {
            throw new $exception;
        }
    }

    /**
     * @param  string $path
     * @return void
     */
    protected function ensureCacheDirectoryExists($path): void
    {
        if (!$this->file->exists(dirname($path))) {
            $this->file->makeDirectory(dirname($path), 0777, true, true);
        }
    }

    /**
     * @param string $key
     * @return string
     */
    protected function path(string $key): string
    {
        $parts = array_slice(str_split($hash = hash('sha256', $key), 2), 0, 2);

        return $this->dataPath . DIRECTORY_SEPARATOR . implode(
            DIRECTORY_SEPARATOR,
            $parts
        ) . DIRECTORY_SEPARATOR . $hash;
    }

    /**
     * @param string $path
     * @return Bucket|null
     */
    protected function getBucket(string $path): ?Bucket
    {
        $bucket = $this->file->get($path);
        if ($bucket) {
            $bucket = unserialize($bucket, ['allowed_classes' => [Node::class, Bucket::class, DateTime::class]]);
        }
        return $bucket;
    }

    /**
     * 获取数据
     *
     * @param string $key
     * @param null   $default
     * @return null|string
     * @throws \Exception
     */
    public function get(string $key, $default = null): ?string
    {
        if (!$this->file->exists($path = $this->path($key))) {
            return $default;
        }
        if (!$bucket = $this->getBucket($path)) {
            return $default;
        }
        if ($node = $bucket->findNode($key)) {
            if ($node->isExpired()) {
                $this->delete($key);
            }
            return $node->getData($default);
        }
        return $default;
    }

    /**
     * 删除一个数据
     *
     * @param string $key
     * @return bool
     */
    public function delete(string $key): bool
    {
        if (!$this->file->exists($path = $this->path($key))) {
            return true;
        }
        if (!($bucket = $this->getBucket($path))) {
            return true;
        }
        $bucket->delNode($key);
        if (!$bucket->isNull()) {
            $this->file->put($path, serialize($bucket));
        }
        return $this->file->delete($path);
    }

    /**
     *  删除所有的数据
     */
    public function flush(): void
    {
        $this->file->deleteDirectory($this->dataPath);
    }

    /**
     * @param string $path
     * @return $this
     */
    public function dataPath(string $path): self
    {
        $this->dataPath = $path;
        return $this;
    }
}


src/Bucket.php
<?php

namespace Yiranzai\File;

use DateTime;

/**
 * Class Bucket
 * @package Yiranzai\Dht
 */
class Bucket
{
    /**
     * @var Node
     */
    protected $head;

    /**
     * Bucket constructor.
     * @param string        $key
     * @param string        $data
     * @param DateTime|null $date
     */
    public function __construct(string $key, string $data = null, DateTime $date = null)
    {
        $this->head = new Node($key, $data, $date);
    }

    /**
     * 链表有几个元素
     *
     * @return int
     */
    public function countNode(): int
    {
        $cur = $this->head;
        $i   = 1;
        while ($cur->next !== null) {
            ++$i;
            $cur = $cur->next;
        }
        return $i;
    }

    /**
     * 增加节点
     *
     * @param string        $key
     * @param string        $data
     * @param DateTime|null $date
     * @return Bucket
     */
    public function addNode(string $key, string $data = null, DateTime $date = null): Bucket
    {
        $cur = $this->head;
        while ($cur->next !== null) {
            $cur = $cur->next;
        }
        $new       = new Node($key, $data, $date);
        $cur->next = $new;
        return $this;
    }

    /**
     * 增加节点
     *
     * @param string        $key
     * @param string        $data
     * @param DateTime|null $date
     * @return Bucket
     */
    public function putNode(string $key, string $data = null, DateTime $date = null): Bucket
    {
        $cur = $this->head;
        if ($cur->key === $key) {
            $cur->data = $data;
            return $this;
        }
        while ($cur->next !== null) {
            $cur = $cur->next;
            if ($cur->key === $key) {
                $cur->data = $data;
                return $this;
            }
        }
        $new       = new Node($key, $data, $date);
        $cur->next = $new;
        return $this;
    }

    /**
     * 紧接着插在$noNum后
     *
     * @param string        $k
     * @param string        $index
     * @param string        $data
     * @param DateTime|null $date
     * @return Bucket
     */
    public function insertNode(string $k, string $index, string $data = null, DateTime $date = null): Bucket
    {
        $cur = $this->head;
        $new = new Node($k, $data, $date);
        if ($cur->key !== $index) {
            while ($cur->next !== null) {
                $cur = $cur->next;
                if ($cur->key === $index) {
                    break;
                }
            }
        }
        $new->next = $cur->next;
        $cur->next = $new;
        return $this;
    }

    /**
     * 删除第$no个节点
     *
     * @param string $key
     * @return Bucket
     */
    public function delNode(string $key): Bucket
    {
        $cur = $this->head;
        if ($cur->key === $key) {
            $this->head = $cur->next;
        }
        while ($cur->next !== null) {
            if ($cur->next->key === $key) {
                $cur->next = $cur->next->next;
                break;
            }
            $cur = $cur->next;
        }
        return $this;
    }

    /**
     * 遍历链表
     *
     * @return Bucket
     * @throws \Exception
     */
    public function showNode(): Bucket
    {
        $cur = $this->head;
        while ($cur->next !== null) {
            $cur = $cur->next;
            echo $cur->getData() . PHP_EOL;
        }
        return $this;
    }

    /**
     * 寻找某节点的值
     *
     * @param string $key
     * @return string|null
     * @throws \Exception
     */
    public function find(string $key): ?string
    {
        $node = $this->findNode($key);
        return $node === null ? null : $node->getData();
    }

    /**
     * 寻找节点
     *
     * @param string $key
     * @return Node|null
     */
    public function findNode(string $key): ?Node
    {
        $cur = $this->head;
        if ($cur->key === $key) {
            return $cur;
        }
        while ($cur->next !== null) {
            $cur = $cur->next;
            if ($cur->key === $key) {
                return $cur;
            }
        }
        return null;
    }

    /**
     * @return bool
     */
    public function isNull(): bool
    {
        return $this->head === null;
    }
}


src/Filesystem.php
<?php

namespace Yiranzai\File;

use FilesystemIterator;

class Filesystem
{

    /**
     * Get the MD5 hash of the file at the given path.
     *
     * @param  string $path
     * @return string
     */
    public function hash($path): string
    {
        return md5_file($path);
    }

    /**
     * Prepend to a file.
     *
     * @param  string $path
     * @param  string $data
     * @return int
     */
    public function prepend($path, $data): int
    {
        if ($this->exists($path)) {
            return $this->put($path, $data . $this->get($path));
        }

        return $this->put($path, $data);
    }

    /**
     * Determine if a file or directory exists.
     *
     * @param  string $path
     * @return bool
     */
    public function exists($path): bool
    {
        return file_exists($path);
    }

    /**
     * Write the contents of a file.
     *
     * @param  string $path
     * @param  string $contents
     * @param  bool   $lock
     * @return int
     */
    public function put($path, $contents, $lock = false): int
    {
        return file_put_contents($path, $contents, $lock ? LOCK_EX : 0);
    }

    /**
     * Get the contents of a file.
     *
     * @param  string $path
     * @param  bool   $lock
     * @return string
     */
    public function get($path, $lock = false): string
    {
        if ($this->isFile($path)) {
            return $lock ? $this->sharedGet($path) : file_get_contents($path);
        }

        throw new \RuntimeException("File does not exist at path {$path}");
    }

    /**
     * Determine if the given path is a file.
     *
     * @param  string $file
     * @return bool
     */
    public function isFile($file): bool
    {
        return is_file($file);
    }

    /**
     * Get contents of a file with shared access.
     *
     * @param  string $path
     * @return string
     */
    public function sharedGet($path): string
    {
        $contents = '';

        $handle = fopen($path, 'rb');

        if ($handle !== false) {
            try {
                if (flock($handle, LOCK_SH)) {
                    clearstatcache(true, $path);

                    $contents = fread($handle, $this->size($path) ?: 1);

                    flock($handle, LOCK_UN);
                }
            } finally {
                fclose($handle);
            }
        }

        return $contents;
    }

    /**
     * Get the file size of a given file.
     *
     * @param  string $path
     * @return int
     */
    public function size($path): int
    {
        return filesize($path);
    }

    /**
     * Append to a file.
     *
     * @param  string $path
     * @param  string $data
     * @return int
     */
    public function append($path, $data): int
    {
        return file_put_contents($path, $data, FILE_APPEND);
    }

    /**
     * Get or set UNIX mode of a file or directory.
     *
     * @param  string $path
     * @param  int    $mode
     * @return mixed
     */
    public function chmod($path, $mode = null)
    {
        if ($mode !== null) {
            return chmod($path, $mode);
        }

        return substr(sprintf('%o', fileperms($path)), -4);
    }

    /**
     * Move a file to a new location.
     *
     * @param  string $path
     * @param  string $target
     * @return bool
     */
    public function move($path, $target): bool
    {
        return rename($path, $target);
    }

    /**
     * Extract the file name from a file path.
     *
     * @param  string $path
     * @return string
     */
    public function name($path): string
    {
        return pathinfo($path, PATHINFO_FILENAME);
    }

    /**
     * Extract the trailing name component from a file path.
     *
     * @param  string $path
     * @return string
     */
    public function basename($path): string
    {
        return pathinfo($path, PATHINFO_BASENAME);
    }

    /**
     * Extract the parent directory from a file path.
     *
     * @param  string $path
     * @return string
     */
    public function dirname($path): string
    {
        return pathinfo($path, PATHINFO_DIRNAME);
    }

    /**
     * Extract the file extension from a file path.
     *
     * @param  string $path
     * @return string
     */
    public function extension($path): string
    {
        return pathinfo($path, PATHINFO_EXTENSION);
    }

    /**
     * Get the file type of a given file.
     *
     * @param  string $path
     * @return string
     */
    public function type($path): string
    {
        return filetype($path);
    }

    /**
     * Get the mime-type of a given file.
     *
     * @param  string $path
     * @return string|false
     */
    public function mimeType($path)
    {
        return finfo_file(finfo_open(FILEINFO_MIME_TYPE), $path);
    }

    /**
     * Get the file's last modification time.
     *
     * @param  string $path
     * @return int
     */
    public function lastModified($path): int
    {
        return filemtime($path);
    }

    /**
     * Determine if the given path is readable.
     *
     * @param  string $path
     * @return bool
     */
    public function isReadable($path): bool
    {
        return is_readable($path);
    }

    /**
     * Determine if the given path is writable.
     *
     * @param  string $path
     * @return bool
     */
    public function isWritable($path): bool
    {
        return is_writable($path);
    }

    /**
     * Move a directory.
     *
     * @param  string $from
     * @param  string $to
     * @param  bool   $overwrite
     * @return bool
     */
    public function moveDirectory($from, $to, $overwrite = false): bool
    {
        if ($overwrite && $this->isDirectory($to) && !$this->deleteDirectory($to)) {
            return false;
        }

        return @rename($from, $to) === true;
    }

    /**
     * Determine if the given path is a directory.
     *
     * @param  string $directory
     * @return bool
     */
    public function isDirectory($directory): bool
    {
        return is_dir($directory);
    }

    /**
     * Recursively delete a directory.
     *
     * The directory itself may be optionally preserved.
     *
     * @param  string $directory
     * @param  bool   $preserve
     * @return bool
     */
    public function deleteDirectory($directory, $preserve = false): bool
    {
        if (!$this->isDirectory($directory)) {
            return false;
        }

        $items = new FilesystemIterator($directory);

        foreach ($items as $item) {
            if ($item->isDir() && !$item->isLink()) {
                $this->deleteDirectory($item->getPathname());
            } else {
                $this->delete($item->getPathname());
            }
        }

        if (!$preserve) {
            /** @scrutinizer ignore-unhandled */ @rmdir($directory);
        }

        return true;
    }

    /**
     * Delete the file at a given path.
     *
     * @param  string|array $paths
     * @return bool
     */
    public function delete($paths): bool
    {
        $paths = is_array($paths) ? $paths : func_get_args();

        $success = true;

        foreach ($paths as $path) {
            try {
                if (!@unlink($path)) {
                    $success = false;
                }
            } catch (\Exception $e) {
                $success = false;
            }
        }

        return $success;
    }

    /**
     * Copy a directory from one location to another.
     *
     * @param  string $directory
     * @param  string $destination
     * @param  int    $options
     * @return bool
     */
    public function copyDirectory($directory, $destination, $options = null): bool
    {
        if (!$this->isDirectory($directory)) {
            return false;
        }

        $options = $options ?: FilesystemIterator::SKIP_DOTS;

        if (!$this->isDirectory($destination)) {
            $this->makeDirectory($destination, 0777, true);
        }

        $items = new FilesystemIterator($directory, $options);

        foreach ($items as $item) {
            $target = $destination . '/' . $item->getBasename();

            if (!$item->isDir()) {
                if (!$this->copy($item->getPathname(), $target)) {
                    return false;
                }
            } else {
                $path = $item->getPathname();

                if (!$this->copyDirectory($path, $target, $options)) {
                    return false;
                }
            }
        }

        return true;
    }

    /**
     * Create a directory.
     *
     * @param  string $path
     * @param  int    $mode
     * @param  bool   $recursive
     * @param  bool   $force
     * @return bool
     */
    public function makeDirectory($path, $mode = 0755, $recursive = false, $force = false): bool
    {
        if ($force) {
            return @mkdir($path, $mode, $recursive);
        }

        return mkdir($path, $mode, $recursive);
    }

    /**
     * Copy a file to a new location.
     *
     * @param  string $path
     * @param  string $target
     * @return bool
     */
    public function copy($path, $target): bool
    {
        return copy($path, $target);
    }

    /**
     * Empty the specified directory of all files and folders.
     *
     * @param  string $directory
     * @return bool
     */
    public function cleanDirectory($directory): bool
    {
        return $this->deleteDirectory($directory, true);
    }

    /**
     * Determine whether the current environment is Windows based.
     *
     * @return bool
     */
    public function windowsOs(): bool
    {
        return stripos(PHP_OS, 'win') === 0;
    }
}
